frontend/catalyst/utils/types.py

```diff
@@ -0,0 +1,60 @@
+# Copyright 2024 Xanadu Quantum Technologies Inc.
+
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+
+#     http://www.apache.org/licenses/LICENSE-2.0
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Tests for type conversion"""
+
+from collections.abc import Sequence
+
+import numpy as np
+from jax._src.lib.mlir import ir
+
+from catalyst.jax_extras import ShapedArray
+
+
+def convert_shaped_arrays_to_tensors(sarrays: Sequence[ShapedArray]):
+    """Convert a sequence of ShapedArrays to a sequence of tensors"""
+    return map(convert_shaped_array_to_tensor, sarrays)
+
+
+def convert_shaped_array_to_tensor(sarray):
+    """Convert an invidual shaped array to an MLIR tensor"""
+    numpy_dtype = sarray.dtype
+    py_shape = sarray.shape
+    mlir_dtype = convert_numpy_dtype_to_mlir(numpy_dtype)
+    return ir.RankedTensorType.get(py_shape, mlir_dtype)
+
+
+# pylint: disable=too-many-return-statements
+def convert_numpy_dtype_to_mlir(dtp):
+    """Convert dtype to MLIR"""
+    if dtp == np.dtype(np.complex128):
+        base = ir.F64Type.get()
+        return ir.ComplexType.get(base)
+    elif dtp == np.dtype(np.complex64):
+        base = ir.F32Type.get()
+        return ir.ComplexType.get(base)
+    elif dtp == np.dtype(np.float64):
+        return ir.F64Type.get()
+    elif dtp == np.dtype(np.float32):
+        return ir.F32Type.get()
+    elif dtp == np.dtype(np.bool_):
+        return ir.IntegerType.get_signless(1)
+    elif dtp == np.dtype(np.int8):
+        return ir.IntegerType.get_signless(8)
+    elif dtp == np.dtype(np.int16):
+        return ir.IntegerType.get_signless(16)
+    elif dtp == np.dtype(np.int32):
+        return ir.IntegerType.get_signless(32)
+    elif dtp == np.dtype(np.int64):
+        return ir.IntegerType.get_signless(64)
+    raise ValueError("Requested type conversion not available.")
```

Comment:
Seems like a map (or 2) would be a better option here. Maybe for a follow-up improvement?

Reply:
https://github.com/PennyLaneAI/catalyst/pull/610/commits/745044a37d02786fcbb18482d26ba4d6683fe962

Reply:
I think I'll have to revert this change. I'm almost 100% sure this causes the following error:

```
FAILED frontend/test/pytest/test_types.py::test_convert_numpy_dtype_to_mlir[inp0-exp0] - AssertionError: assert ComplexType(complex<f64>) == ComplexType(complex<f64>)
 +  where ComplexType(complex<f64>) = convert_numpy_dtype_to_mlir(dtype('complex128'))
FAILED frontend/test/pytest/test_types.py::test_convert_numpy_dtype_to_mlir[inp1-exp1] - AssertionError: assert ComplexType(complex<f32>) == ComplexType(complex<f32>)
 +  where ComplexType(complex<f32>) = convert_numpy_dtype_to_mlir(dtype('complex64'))
FAILED frontend/test/pytest/test_types.py::test_convert_numpy_dtype_to_mlir[inp2-exp2] - AssertionError: assert F64Type(f64) == F64Type(f64)
 +  where F64Type(f64) = convert_numpy_dtype_to_mlir(dtype('float64'))
FAILED frontend/test/pytest/test_types.py::test_convert_numpy_dtype_to_mlir[inp3-exp3] - AssertionError: assert F32Type(f32) == F32Type(f32)
 +  where F32Type(f32) = convert_numpy_dtype_to_mlir(dtype('float32'))
FAILED frontend/test/pytest/test_types.py::test_convert_numpy_dtype_to_mlir[inp4-exp4] - AssertionError: assert IntegerType(i1) == IntegerType(i1)
 +  where IntegerType(i1) = convert_numpy_dtype_to_mlir(dtype('bool'))
FAILED frontend/test/pytest/test_types.py::test_convert_numpy_dtype_to_mlir[inp5-exp5] - AssertionError: assert IntegerType(i8) == IntegerType(i8)
 +  where IntegerType(i8) = convert_numpy_dtype_to_mlir(dtype('int8'))
FAILED frontend/test/pytest/test_types.py::test_convert_numpy_dtype_to_mlir[inp6-exp6] - AssertionError: assert IntegerType(i16) == IntegerType(i16)
 +  where IntegerType(i16) = convert_numpy_dtype_to_mlir(dtype('int16'))
FAILED frontend/test/pytest/test_types.py::test_convert_numpy_dtype_to_mlir[inp7-exp7] - AssertionError: assert IntegerType(i32) == IntegerType(i32)
 +  where IntegerType(i32) = convert_numpy_dtype_to_mlir(dtype('int32'))
FAILED frontend/test/pytest/test_types.py::test_convert_numpy_dtype_to_mlir[inp8-exp8] - AssertionError: assert IntegerType(i64) == IntegerType(i64)
 +  where IntegerType(i64) = convert_numpy_dtype_to_mlir(dtype('int64'))
 ```
 
 Just because the contexts are different. Essentially, I believe we cannot precompute the types.